upstream: load distribution in Total Panic mode

cpakulski opened the pull request:
Description: Changed how load is calculated when all priority levels are in panic mode. Each priority level receives percentage of the traffic related to the number of hosts in that priority regardless of the health status of the hosts. This smooths out how traffic is shifted when hosts become unhealthy. See #4685 for design proposal and discussion.

Risk Level: Medium

Testing: Unit tests.

Docs Changes: Yes - updated section about panic threshold.

Release Notes: N/A
Fixes #4685

snowp commented on docs/root/intro/arch_overview/upstream/load_balancing/panic_threshold.rst:
```diff
@@ -79,8 +79,16 @@ priority.
-If all hosts become unhealthy normalized total health is 0%, and if the panic threshold is above 0%
-all traffic will be redirected to P=0.
+If the number of healthy hosts keeps decreasing, individual priority levels will start entering panic mode.
+Until at least one priority level is not in panic mode, load distribution is calculated as it is 
```
I think flipping this sentence around would read a bit better: `Load distribution is calculated as describe above as long as there are priority levels not in panic mode`.

cpakulski replied:
Done.

snowp commented on docs/root/intro/arch_overview/upstream/load_balancing/panic_threshold.rst:
```diff
@@ -79,8 +79,16 @@ priority.
+If the number of healthy hosts keeps decreasing, individual priority levels will start entering panic mode.
+Until at least one priority level is not in panic mode, load distribution is calculated as it is 
+described above. When all priority levels enter the panic mode, load calculation algorithm changes.
+In such state each priority level receives percentage of the traffic which represents the number
```
`In this case`

`receives traffic relative to the number of hosts in that priority level`

cpakulski replied:
Done.

snowp commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
+  // This is just a placeholder, because no traffic will be directed to
+  // priority 0 anyways, because it is empty.
+  if (panic_threshold == 0 && normalized_total_availability == 0) {
+    per_priority_load_.healthy_priority_load_.get()[0] = 100;
```
How come this wasn't necessary previously?

cpakulski replied:
I get segfault in unit tests for corner cases, like when all priorities are completely empty. Previously it was not a problem, but now the total number of hosts in used in division, so if it is zero, I get segfault (div by zero).

snowp commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
     // healthy hosts.
     const HostSet& priority_host_set = *priority_set_.hostSetsPerPriority()[i];
     per_priority_panic_[i] =
         (normalized_total_availability == 100 ? false : isGlobalPanic(priority_host_set));
```
does the function name `isGlobalPanic` still make sense? It seems a bit confusing to have both "total panic" and "global panic" - maybe we should rename the function?

cpakulski replied:
Good point. It was confusing for me too, but I got used to it. Renamed to _isHostSetInPanic_.

snowp commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
+  uint32_t total_hosts_num = 0;
+  for (size_t i = 0; i < per_priority_panic_.size(); i++) {
+    const HostSet& host_set = *priority_set_.hostSetsPerPriority()[i];
+    total_hosts_num += host_set.hosts().size() - host_set.excludedHosts().size();
```
I think `excludedHosts` are routed to in panic, so we probably want to include them here? I could be wrong, would be good to verify

cpakulski replied:
I have to verify it.

cpakulski replied:
Done. excludedHosts are not used when calculating load distribution in TotalPanic.

snowp commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
+  // first calculate total number of hosts across all priorities regardless
+  // whether they are healthy or not.
+  uint32_t total_hosts_num = 0;
+  for (size_t i = 0; i < per_priority_panic_.size(); i++) {
```
you can use `std::accumulate` here

cpakulski replied:
Done. Good idea. Thanks!

snowp commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
+  // This is corner case when panic is disabled and there is no hosts available.
+  // But the load must be distributed somewhere.
+  // This is just a placeholder, because no traffic will be directed to
+  // priority 0 anyways, because it is empty.
```
technically its either empty or has no healthy hosts, right? would be good to be specific

alyssawilk replied:
Yeah, this comment doesn't seem quite right.  P[0] may have hosts (it's not necessarily empty).  In theory it would just fail to accept queries right?

alyssawilk replied:
Also I think there's a comment above recalculatePerPriorityState which I think may need updating.

cpakulski replied:
Updated the comment above recalculatePerPriorityState. Added explanation why 100% of load is assigned to P=0 when there are no healthy hosts and panic is disabled. Other routines expect that the sum of all percentages adds to 100%. If this is not the case there are assertions which fail. When panic is disabled, the load balancers will look for healthy hosts and because there is none, error will be returned and nothing will be routed to backend cluster.

snowp commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
+// of healthy hosts in the priority, but based on number of hosts
+// in each priority regardless of its health.
+void LoadBalancerBase::recalculateLoadInTotalPanic() {
+  // first calculate total number of hosts across all priorities regardless
```
nit: proper grammar in comments (capitalize `first`)

cpakulski replied:
Done.

snowp commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
+
+  // Now iterate through all priority levels and calculate how much
+  // load is supposed to go to each priority. In panic mode the calculation
+  // is based n ot on the number of healthy hosts but based on the number of
```
`n ot` -> `not`

cpakulski replied:
Done.

snowp commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
+  uint32_t total_load = 100;
+  for (size_t i = 0; i < per_priority_panic_.size(); i++) {
+    const HostSet& host_set = *priority_set_.hostSetsPerPriority()[i];
+    auto hosts_num = host_set.hosts().size() - host_set.excludedHosts().size();
```
nit: `const` where possbile (here and next line)

cpakulski replied:
Done.

snowp commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
+    per_priority_load_.degraded_priority_load_.get()[i] = 0;
+    total_load -= priority_load;
+  }
+  // add to priority zero whatever if left
```
nit: proper casing + period

cpakulski replied:
Done.

snowp commented on test/common/upstream/load_balancer_impl_test.cc:
```diff
@@ -130,8 +130,18 @@ TEST_P(LoadBalancerBaseTest, PrioritySelection) {
   // unhealthy primary host.
-  EXPECT_EQ(100, lb_.percentageLoad(0));
-  EXPECT_EQ(0, lb_.percentageLoad(1));
+  // primary and failover are in panic mode. Load distribution is based
```
nit: `Primary`

cpakulski replied:
Done.

snowp commented on test/common/upstream/load_balancer_impl_test.cc:
```diff
@@ -130,8 +130,18 @@ TEST_P(LoadBalancerBaseTest, PrioritySelection) {
   EXPECT_CALL(context, determinePriorityLoad(_, _)).WillRepeatedly(ReturnRef(priority_load));
   // With both the primary and failover hosts unhealthy, we should select an
```
this comment is out of date now

cpakulski replied:
Done.

snowp commented on test/common/upstream/load_balancer_impl_test.cc:
```diff
@@ -130,8 +130,18 @@ TEST_P(LoadBalancerBaseTest, PrioritySelection) {
+
+  // Modify number of hosts in failover, but leave them in unhealthy state
+  // primary and secondary are in panic mode, so load distribution is
+  // based on number of host regardless on their health.
```
`regardless of their health`

cpakulski replied:
Done.

snowp commented on test/common/upstream/load_balancer_impl_test.cc:
```diff
@@ -130,8 +130,18 @@ TEST_P(LoadBalancerBaseTest, PrioritySelection) {
+  EXPECT_EQ(50, lb_.percentageLoad(1));
+  EXPECT_EQ(&host_set_, &lb_.chooseHostSet(&context).first);
+
+  // Modify number of hosts in failover, but leave them in unhealthy state
```
`in an/the unhealthy state`

cpakulski replied:
Done.

snowp commented on test/common/upstream/load_balancer_impl_test.cc:
```diff
@@ -403,17 +415,24 @@ TEST_P(LoadBalancerBaseTest, GentleFailoverWithExtraLevels) {
   // All levels are completely down. 100% of traffic should go to P=0
```
update comment

cpakulski replied:
Done. Comment updated with new logic.

snowp commented on test/common/upstream/load_balancer_impl_test.cc:
```diff
@@ -403,17 +415,24 @@ TEST_P(LoadBalancerBaseTest, GentleFailoverWithExtraLevels) {
+  ASSERT_THAT(getPanic(), ElementsAre(true, true, true));
 
   // Rounding errors should be picked up by the first healthy priority.
+  // This situation is called TotalPanic, as all priority levels are in panic
```
its a bit odd that this explanation appears here when the above example is also in total panic, maybe move it up?

cpakulski replied:
Added the same explanation to all cases when TotalPanic is used.

snowp commented on test/common/upstream/load_balancer_impl_test.cc:
```diff
@@ -427,13 +446,32 @@ TEST_P(LoadBalancerBaseTest, GentleFailoverWithExtraLevels) {
 
   // Rounding errors should be picked up by the first priority with degraded hosts when
   // there are no healthy priorities.
+  // Just disable panic threshold. Otherwise TotalPanic calculation
```
minor nit but I would rephrase this as `Disable panic threshold to prevent total panic from kicking in`. Alternatively, tweak the parameters of the test so that we're still able to test the behavior of rounding when everything is degraded without having to disable panic.

cpakulski replied:
Done.

cpakulski commented:
@snowp Thank you for code review. I am working on addressing your comments. There is one corner case which popped up during unit testing. This is when there is no heathy hosts across all priority levels. After my latest changes the algorithm just keeps running in total panic mode and distributes load to all priorities. However, there are some assumptions in the existing code that if there are no healthy hosts, chooseHost method should return nullptr. Interestingly, not all LBs have such assumptions. Ring LB has and its unit tests fail. Also aggregate cluster unit tests also assume that chooseHost returns nullptr. I think that TotalPanic mode should apply only when all levels are in panic mode and there is at least one healthy host in the cluster. I will change my code to work that way and will push code again.

snowp commented:
That seems like an odd behavior, I would have expected panic mode to trigger even when there are no healthy hosts. Otherwise you might be saved by panic mode, only for your traffic to start failing once all hosts go unhealthy. @mattklein123 do you know the what the original intended behavior was around panic routing /w no healthy upstreams?

mattklein123 commented:
> That seems like an odd behavior, I would have expected panic mode to trigger even when there are no healthy hosts. Otherwise you might be saved by panic mode, only for your traffic to start failing once all hosts go unhealthy. @mattklein123 do you know the what the original intended behavior was around panic routing /w no healthy upstreams? 

Definitely a bug. Please fix!

cpakulski commented:
Thanks for clarification! Will adjust the code accordingly.

cpakulski commented:
wip: ping

cpakulski commented:
/retest

cpakulski commented:
@snowp I addressed all your code review comments accept https://github.com/envoyproxy/envoy/pull/9343#discussion_r357990021. I have to check the code to understand how a backend node ends up in `excludedHost` list and if we should route the traffic there while in panic mode.

cpakulski commented:
@snowp I addressed the last issue https://github.com/envoyproxy/envoy/pull/9343#discussion_r357990021. `excludedHosts` array is filled in https://github.com/envoyproxy/envoy/blob/v1.12.2/source/common/upstream/upstream_impl.cc#L855-L872 and traffic is router to `excludedHosts` in TotalPanic state. The code is ready for another pass.

snowp commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -238,22 +238,68 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
+void LoadBalancerBase::recalculateLoadInTotalPanic() {
+  // First calculate total number of hosts across all priorities regardless
+  // whether they are healthy or not.
+  const uint32_t total_hosts_num = std::accumulate(
```
nit: i think i'd prefer `total_hosts_count`

cpakulski replied:
Done.

snowp commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -238,22 +238,68 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
+    per_priority_load_.degraded_priority_load_.get()[i] = 0;
+    total_load -= priority_load;
+  }
+  // Add to priority zero whatever if left.
```
Do we need to worry about whether P0 has any hosts? Should we assign the rounding error to the first non-empty priority?

cpakulski replied:
That would require configuration with empty P0 and not empty P>0. To be on the safe side I added loop to find the first not empty priority. Thanks for suggestion.

snowp commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -238,22 +238,68 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
+    const HostSet& host_set = *priority_set_.hostSetsPerPriority()[i];
+    auto hosts_num = host_set.hosts().size() - host_set.excludedHosts().size();
+
+    uint32_t priority_load = 100 * hosts_num / total_hosts_num;
```
nit: const

cpakulski replied:
Done.

snowp commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -238,22 +238,68 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
+  uint32_t total_load = 100;
+  for (size_t i = 0; i < per_priority_panic_.size(); i++) {
+    const HostSet& host_set = *priority_set_.hostSetsPerPriority()[i];
+    auto hosts_num = host_set.hosts().size() - host_set.excludedHosts().size();
```
where did we land on whether to include excluded hosts? if they are used during panic routing i think it would make sense to not subtract them here?

cpakulski replied:
Done.

cpakulski replied:
I found out that excluded hosts are those for which healthcheck is not resolved yet. I understand that this was added to avoid entering panic mode when bunch of new hosts is added for which health is not known yet. Whether to route to them in panic mode? I guess the answer is yes and that is how it is implemented now.

The existence of `host_set.excludedHosts().size();` here is a bug, because I removed it from calculation of the total load in my previous commit but I did not remove it here when hosts are calculated for each priority. I was perplexed why none of tests caught it. So I added a new test to include excludedHosts case, but the behavior was not what I expected. It happened that there was a bug in test setup routine. The following line is wrong: https://github.com/envoyproxy/envoy/blob/v1.12.2/test/common/upstream/load_balancer_impl_test.cc#L87. It was probably copy-pasted.

cpakulski commented:
/retest

snowp commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -238,21 +238,74 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
+  uint32_t total_load = 100;
+  for (size_t i = 0; i < per_priority_panic_.size(); i++) {
+    const HostSet& host_set = *priority_set_.hostSetsPerPriority()[i];
+    const auto hosts_num = host_set.hosts().size();
```
nit: since this isnt reused i would just inline it

cpakulski replied:
I am reusing it now for the logic related to rounding errors.

snowp commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -238,21 +238,74 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
+  }
+
+  // Find the first priority which is not empty and add remaining load.
+  for (size_t i = 0; i < per_priority_panic_.size(); i++) {
```
in other places we do this kind of computation we determine what priority to allocate the remainder to during the loop where we distribute load, see https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/load_balancer_impl.cc#L205

i'd probably just keep a `size_t first_nonempty` variable that gets set to the index of the first priority with  hosts in it and then use that to allocate the remainder.

cpakulski replied:
Done.

snowp commented on test/common/upstream/load_balancer_impl_test.cc:
```diff
@@ -402,21 +426,34 @@ TEST_P(LoadBalancerBaseTest, GentleFailoverWithExtraLevels) {
+  ASSERT_THAT(getPanic(), ElementsAre(true, true, true));
 
-  // Rounding errors should be picked up by the first healthy priority.
+  // Rounding errors should be picked up by the first priority.
```
Can you add a test for when the priority that gets given the remainder is not the first priority?

cpakulski replied:
Done.

snowp commented on test/extensions/clusters/aggregate/cluster_test.cc:
```diff
@@ -217,7 +217,16 @@ TEST_F(AggregateClusterTest, AllHostAreUnhealthyTest) {
   EXPECT_CALL(primary_load_balancer_, chooseHost(_)).WillRepeatedly(Return(host));
   EXPECT_CALL(secondary_load_balancer_, chooseHost(_)).WillRepeatedly(Return(nullptr));
 
-  for (int i = 0; i < 100; ++i) {
```
mind including a comment here about whats going on? its not super obvious from reading the code

cpakulski replied:
Comments were added.

cpakulski commented:
/retest

alyssawilk commented on docs/root/intro/version_history.rst:
```diff
@@ -60,6 +60,7 @@ Version history
 * tracing: added initial support for AWS X-Ray (local sampling rules only) :ref:`X-Ray Tracing <envoy_api_msg_config.trace.v2alpha.XRayConfig>`.
 * tracing: added tags for gRPC request path, authority, content-type and timeout.
 * udp: added initial support for :ref:`UDP proxy <config_udp_listener_filters_udp_proxy>`
+* upstream: changed load distribution algorithm when all priorities enter :ref:`panic mode<arch_overview_load_balancing_panic_threshold>`. 
```
This will need to be moved up to the 1.4 section since we just cut the quarterly release :-)

cpakulski replied:
Done.

alyssawilk commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -58,7 +58,8 @@ class LoadBalancerBase : public LoadBalancer {
    */
-  bool isGlobalPanic(const HostSet& host_set);
+  bool isHostSetInPanic(const HostSet& host_set);
+  void recalculateLoadInTotalPanic();
```
Comment please

cpakulski replied:
Done.

cpakulski commented:
/retest

snowp commented on docs/root/intro/version_history.rst:
```diff
@@ -66,6 +67,7 @@ Version history
 * tracing: added initial support for AWS X-Ray (local sampling rules only) :ref:`X-Ray Tracing <envoy_api_msg_config.trace.v2alpha.XRayConfig>`.
 * tracing: added tags for gRPC request path, authority, content-type and timeout.
 * udp: added initial support for :ref:`UDP proxy <config_udp_listener_filters_udp_proxy>`
+* fault: fixed an issue where the http fault filter would repeatedly check the percentage of abort/delay when the `x-envoy-downstream-service-cluster` header was included in the request to ensure that the actual percentage of abort/delay matches the configuration of the filter.
```
bad merge?

cpakulski replied:
Just corrected it and pushed again. Not sure how it happened. I was getting merge conflict for that file and probably hit wrong keys in vim. Thanks for spotting it.
